Sources/TensorFlow/Layers/Recurrent.swift

```diff
@@ -420,12 +425,150 @@ public struct RecurrentLayer<Cell: RecurrentLayerCell>: Layer {
+  @differentiable
+  public static func sum(_ lhs: Tensor, _ rhs: Tensor) -> Tensor {
+    lhs + rhs
+  }
+}
+
+@differentiable
+public func sum<T: Mergeable>(
+  _ first: T,
+  _ second: T
+) -> T {
+  T.sum(first, second)
+}
+
+public struct BidirectionalRecurrentLayer<Cell: RecurrentLayerCell>: Layer
+where Cell.TimeStepOutput: Mergeable {
+  public typealias Input = [Cell.TimeStepInput]
+  public typealias Output = [Cell.TimeStepOutput]
+  public typealias MergeFunction = @differentiable (Cell.TimeStepOutput, Cell.TimeStepOutput) -> Cell.TimeStepOutput
+
+  /// A wrapper around a `@differentiable` merge function.
+  ///
+  /// - Note: this exists as a workaround for runtime crashes regarding `@differentiable`function
+  ///   stored properties (TF-1122).
+  private class _MergeFunction {
+    var function: MergeFunction
+    init(_ function: @escaping MergeFunction) {
+      self.function = function
+    }
+  }
+
+  @noDerivative private let _mergeFunction: _MergeFunction
+  /// The forward recurrent layer.
+  public var forward: RecurrentLayer<Cell>
+  /// The backward recurrent layer.
+  public var backward: RecurrentLayer<Cell>
+  /// The differentiable function used for merging forward and backward recurrent layer outputs.
+  @noDerivative public var mergeFunction: MergeFunction {
+    _mergeFunction.function
+  }
+
+  /// Creates an instance from the given recurrent layer cell and merge function.
+  ///
+  /// - Note: The default merge mode in Keras is `concat`, which is not currently supported.
+  ///   When the `concat` mode is added, it can be used as the default value.
+  public init(_ cell: @autoclosure () -> Cell, mergeFunction: @escaping MergeFunction) {
+    forward = RecurrentLayer(cell())
+    backward = RecurrentLayer(cell())
+    _mergeFunction = .init(mergeFunction)
+  }
+
+  @differentiable
+  public func callAsFunction(
+    _ inputs: Input,
+    initialForwardLayerState: Cell.State,
+    initialBackwardLayerState: Cell.State
+  ) -> Output {
+    let forwardOutputs = forward(
+      inputs, initialState: initialForwardLayerState)
+    
+    // TODO: Replace with inputs.reversed() after it become differentiable.
+    var inputsReversed = Input()
+    
+    for forwardIndex in 0 ..< withoutDerivative(at: inputs.count) {
+        let backwardIndex = withoutDerivative(at: inputs.count - 1 - forwardIndex)
+        inputsReversed.append(inputs[backwardIndex])
+    }
+    
+    let backwardOutputs = backward(
+        inputsReversed, initialState: initialBackwardLayerState)
+
+    var outputs = Output()
+
+    for forwardIndex in  0 ..< withoutDerivative(at: inputs.count) {
+        let backwardIndex = withoutDerivative(at: inputs.count - 1 - forwardIndex)
+        outputs.append(mergeFunction(forwardOutputs[forwardIndex], backwardOutputs[backwardIndex]))
+    }
+
+    return outputs
+  }
+
+  @differentiable
+  public func callAsFunction(_ inputs: Input) -> Output {
```

Comment:
Please add a precondition:
```swift
    precondition(!inputs.isEmpty, "'inputs' must be non-empty.")
```

Same for the `lastOutput` methods below.

Reply:
Added.